Use the post-type-cast version of the attribute to validate numericality

repinel opened the pull request:
This should fix the issue #12134 as started by @trev in #18617 and with the changed indicated by @sgrif.

When validating numeric attributes, the actual value was not being used since the validator was parsing it internally to `float`.

egilburg commented on activemodel/lib/active_model/validations/numericality.rb:
```diff
@@ -64,14 +62,16 @@ def validate_each(record, attr_name, value)
-      def parse_raw_value_as_an_integer(raw_value)
-        raw_value.to_i if raw_value.to_s =~ /\A[+-]?\d+\z/
+      def is_integer?(raw_value)
+        parsed_value = raw_value.to_i if raw_value.to_s =~ /\A[+-]?\d+\z/
```
`to_i` will always be `present?` (whether 0 or otherwise) so this can be simplified to 

```ruby
match = raw_value.to_s =~ /\A[+-]?\d+\z/
match.present?
```

repinel commented:
@egilburg good point. I'll make the appropriate changes.

trev commented:
:+1:

egilburg commented:
@trev `=~` will return `nil`/`Integer`. If you want predicate, use `===`

trev commented:
@egilburg I mean the whole commit, not your suggestion :)

trev commented:
Nevermind! It's too early, I completely overlooked the tests. Carry on.

repinel committed:
Conditionally convert the raw_value received by the numeric validator.

This fixes the issue where you may be comparing (using a numeric
validator such as `greater_than`) numbers of a specific Numeric type
such as `BigDecimal`.

Previous behavior took the numeric value to be validated and
unconditionally converted to Float. For example, due to floating point
precision, this can cause issues when comparing a Float to a BigDecimal.

Consider the following:

```
    validates :sub_total, numericality: {
      greater_than: BigDecimal('97.18')
    }
```

If the `:sub_total` value BigDecimal.new('97.18') was validated against
the above, the following would be valid since `:sub_total` is converted
to a Float regardless of its original type. The result therefore becomes
Kernel.Float(97.18) > BigDecimal.new('97.18')

The above illustrated behavior is corrected with this patch by
conditionally converting the value to validate to float.

Use the post-type-cast version of the attribute to validate numericality

[Roque Pinel & Trevor Wistaff]
